homeassistant/components/axis/switch.py

```diff
@@ -1,46 +1,88 @@
 """Support for Axis switches."""
 
+from collections.abc import Callable, Iterable
+from dataclasses import dataclass
+from functools import partial
 from typing import Any
 
 from axis.models.event import Event, EventOperation, EventTopic
 
-from homeassistant.components.switch import SwitchEntity
+from homeassistant.components.switch import (
+    SwitchDeviceClass,
+    SwitchEntity,
+    SwitchEntityDescription,
+)
 from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EntityCategory
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 
 from .entity import AxisEventEntity
 from .hub import AxisHub
 
 
+@dataclass(frozen=True, kw_only=True)
+class AxisSwitchDescription(SwitchEntityDescription):
+    """Axis switch entity description."""
+
+    event_topic: EventTopic
+    """Event topic that provides state updates."""
+    name_fn: Callable[[AxisHub, Event], str]
+    """Function providing the corresponding name to the event ID."""
+    supported_fn: Callable[[AxisHub, Event], bool]
+    """Function validating if event is supported."""
+
+
+ENTITY_DESCRIPTIONS = (
+    AxisSwitchDescription(
+        key="Relay state control",
+        device_class=SwitchDeviceClass.OUTLET,
+        entity_category=EntityCategory.CONFIG,
+        event_topic=EventTopic.RELAY,
+        supported_fn=lambda hub, event: isinstance(int(event.id), int),
+        name_fn=lambda hub, event: hub.api.vapix.ports[event.id].name,
+    ),
+)
+
+
 async def async_setup_entry(
     hass: HomeAssistant,
     config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
-    """Set up a Axis switch."""
+    """Set up the Axis switch platform."""
     hub = AxisHub.get_hub(hass, config_entry)
 
     @callback
-    def async_create_entity(event: Event) -> None:
-        """Create Axis switch entity."""
-        async_add_entities([AxisSwitch(event, hub)])
+    def register_platform(descriptions: Iterable[AxisSwitchDescription]) -> None:
+        """Register entity platform to create entities on event initialized signal."""
+
+        @callback
+        def create_entity(description: AxisSwitchDescription, event: Event) -> None:
+            """Create Axis entity."""
+            if description.supported_fn(hub, event):
+                async_add_entities([AxisSwitch(hub, description, event)])
+
+        for description in descriptions:
+            hub.api.event.subscribe(
+                partial(create_entity, description),
```

Comment:
Why is event a parameter if you're not using it in the partial?

Reply:
The partial is prepare the callback as the axis library doesn't expect the description to be provided as input. This is based on the UniFi integration and is an improvement provided by bdraco :)